File: 57. Insert Interval.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        i = 0
        res = []
        n = len(intervals)
        intr = newInterval
        while i < n and intervals[i].end < newInterval.start:
            res.append(intervals[i])
            i += 1

        while i < n and intervals[i].start <= newInterval.end:
            intr = Interval(min(intervals[i].start, intr.start), max(intervals[i].end, intr.end))
            i += 1
        res.append(intr)

        while i < n:
            res.append(intervals[i]);
            i += 1

        return res

File: 57. Insert Interval.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    public List<Interval> insert(List<Interval> intervals, Interval newInterval) {
        List<Interval> res = new ArrayList<>();
        if (intervals.size() == 0) {
            res.add(newInterval);
            return res;
        }
        int newStart = -1;
        int newEnd = -1;
        int p = newInterval.start, q = newInterval.end;
        for (Interval itvl : intervals) {
            if (newStart == -1) {
                if (itvl.end < p) {
                    res.add(new Interval(itvl.start, itvl.end));
                } else if(itvl.start > q) {
                    res.add(newInterval);
                    newStart = p;
                    newEnd = q;
                    res.add(new Interval(itvl.start, itvl.end));
                } else {
                    newStart = itvl.start <= p ? itvl.start : p;
                    if (itvl.end >= q) {
                        newEnd = itvl.end;
                        res.add(new Interval(newStart, newEnd));
                    }
                }
            } else if (newEnd == -1) {
                if (itvl.start > q) {
                    newEnd = q;
                    res.add(new Interval(newStart, newEnd));
                    res.add(new Interval(itvl.start, itvl.end));
                } else if (itvl.end >= q) {
                    newEnd = itvl.end;
                    res.add(new Interval(newStart, newEnd));
                }
            } else {
                res.add(new Interval(itvl.start, itvl.end));
            }
        }
        if (newStart == -1) {
            res.add(new Interval(p, q));
        } else if (newEnd == -1) {
            res.add(new Interval(newStart, Math.max(q, intervals.get(intervals.size() - 1).end)));
        }
        return res;
    }
}